policy: policy calculation enhancements, factoring out of policy generation from endpoint package

ianvernon opened the pull request:
Based on ~~#6275~~ .  I'm waiting for that to get merged before merging this. I rebased on top of it to avoid conflicts. 

This PR is part of the overall work to improve the performance / scalability of policy calculation within Cilium. In order to make testing scalability and performance of policy easier, it behooves us to factor out policy generation for a given set of labels (i.e., the security-relevant labels for an endpoint which comprise its security identity) into in `pkg/policy`. 

Add a new type which encompasses the policy at L3, L4, and L7 that is generated based off of the provided set of labels. This currently wraps the pre-existing types that were distinct fields within an `Endpoint`, `L3Policy`, `{Desired,Realized}L4Policy`, and `{desired,realized}PolicyMapState`. This is done so that in the future, we can hide how policy is plumbed from the endpoint structure itself, allowing for an iterative refactor of the internals of this new type to improve performance of policy regeneration.

This PR improves policy calculation performance as well. When the initial pass is performed over the rules to determine whether ingress and egress policy enforcement apply for the provided set of labels, the list of rules which select the endpoint itself is also returned. By returning this subset of the rules in the policy repository, when we pass over the rules to generate policy for each type of policy object generated (`CIDRPolicy`, `L4Policy`, `MapState`), we no longer have to check whether the rules being analyzed select the given set of labels. Caching the rules that match locally improves performance, because performing matching operations against `EndpointSelector` is costly; this reduces the number of times matching is performed against the `EndpointSelector` to be done only once during policy calculation. It also means that policy generation is only performed against the subset of rules which select the set of labels. In the case in which only one rule selects the set of labels out of thousands of rules, not only is only that rule considered when the policy rules are translated, we no longer perform matching operations against each and every one of those rules which don't even select the endpoint multiple times.

We track whether or not to check whether the EndpointSelector in the rule matches the set of provided labels by providing a new field in the `SearchContext`, called `rulesSelect`. If set, the code does not check whether the `EndpointSelector` selects the given set of labels, allowing matching to be skipped.

I ran benchmarks against `regeneratePolicy`, and found that this leads to around 54 % decrease in amount of time taken to regenerate policy in the case where no rules select the endpoint for which policy is being regenerated:

Previous:
```
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=false ipv4= ipv6= k8sPodName=/ policyChanged=true policyRegenerationTime=562.150107ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyChanged=false policyRegenerationTime=451.340492ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyChanged=false policyRegenerationTime=696.477095ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyChanged=false policyRegenerationTime=661.700047ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyChanged=false policyRegenerationTime=567.049182ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyChanged=false policyRegenerationTime=465.258572ms subsys=endpoint
PASS: policy_test.go:215: EndpointSuite.BenchmarkRegeneratePolicyRules	       5	 568447601 ns/op
```

With changes in this PR:

```
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=false ipv4= ipv6= k8sPodName=/ policyRegenerationTime=197.632603ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=205.555571ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=222.051934ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=441.657516ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=203.362918ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=206.866816ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=415.307384ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=206.394149ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=227.010221ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=317.161773ms subsys=endpoint
level=info msg="Completed endpoint policy recalculation" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=256 forcedRegeneration=true ipv4= ipv6= k8sPodName=/ policyRegenerationTime=201.965828ms subsys=endpoint
PASS: policy_test.go:215: EndpointSuite.BenchmarkRegeneratePolicyRules	      10	 264943078 ns/op
OK: 1 passed
```

Things this PR does not do:

* Change the Endpoint structure itself. Ideally we can package this new `Policy` type as a field within the endpoint. I'd like to keep that to a future PR to avoid having too much going on in this PR. 
* Unify policy generation. The same structures as before (`L4Policy`, `CIDRPolicy`, BPF PolicyMap state) are kept the same as before. Changing of these structures will be done by future PRs. 
* Add unit tests for checking the state of a `Policy` (yet). I didn't want to add a ton of tests for something which would be changed. FWIW, the results of it are only used in `Endpoint`, and the CI has passed, which was the main test for `regeneratePolicy` before. I can add these in the future / as follow-up commits if people agree on the general change I'm making here.

Signed-off by: Ian Vernon <ian@cilium.io>

```release-note
Increase performance of policy computation when rules don't select labels by approximately 50 %
```

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6270)
<!-- Reviewable:end -->

Commit:
policy: add new Policy type

This groups separate types that encode resolved policy, e.g. L4Policy,
CIDRPolicy, into one structure.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add new ResolvePolicy function

This function returns a generated policy object based off the following state:

* Set of labels for which the policy is computed.
* Set of all identities.
* List of rules in the policy repository.

The first step of this new function is to figure out whether ingress or egress
enforcement applies to the provided set of labels. All rules need to be analyzed
as to whether they match these labels. While this check is done, cache the rules
that match locally. This increases performance of policy evaluation by reducing
the number of rules which are iterated over when calculating policy at each
protocol layer for an endpoint, as well as performing the matching operation
upon the endpoint selector of the rules only once, as opposed to against all
rules in the repository for each protocol layer.

Since the list of rules are already predetermined to select the set of labels,
the check as to whether the rules select the set of labels does not need to be
performed again. This allows for setting of the `rulesSelect` flag in the
SearchContext objects for both ingress and egress policy generation, which
skips said aforementioned check in the subsequent policy generation at L4, for
CIDRs, etc.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add rulesSelect field to SearchContext

rulesSelect specifies whether or not to check whether a rule which is
being analyzed against its SearchContext matches either the From or To
fields within the SearchContext. This is used to avoid using
EndpointSelector.Matches() while analyzing rules if possible, since
said function is costly in terms of performance. Currently, this flag
is always false, so no functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: convert L4Policy to Key

This allows for the L4Policy to be converted to a representation the datapath
can understand. In the future, this will be done elsewhere, but in order to
unify policy generation, it has been moved into the policy package.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: plumb L3-only PolicyKey calculation into ResolvePolicy

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: use policy.ResolvePolicy to determine desired policy state

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: remove ComputePolicyEnforcement tests

This functionality is tested in the policy package in
pkg/policy/repository_test.go:TestComputePolicyEnforcementAndRules.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove unused code

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/policy/resolve.go:
```diff
@@ -0,0 +1,25 @@
+
+// Policy is a structure which contains the resolved policy across all layers
+// (L3, L4, and L7).
+type Policy struct {
```
@aanm suggested we call this an `EndpointPolicy`, since it is currently the policy for a given endpoint. Thoughts?

ianvernon commented on pkg/policy/resolve.go:
```diff
@@ -21,5 +21,5 @@ type Policy struct {
 	CIDRPolicy           *CIDRPolicy
```
I will update this commit message to be more accurate.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -115,6 +117,28 @@ func (l4 *L4Filter) AllowsAllAtL3() bool {
 	return l4.Endpoints.SelectsAllEndpoints()
 }
 
+// ToPolicyMapKeys converts filter into a list of PolicyKeys.
```
update documentation

Commit:
policy: rename Policy to EndpointPolicy

This type is generated on a per-endpoint basis, so name it as such.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -0,0 +1,124 @@
+	return identities
+}
+
+func (p *EndpointPolicy) computeDesiredL4PolicyMapEntries(identityCache cache.IdentityCache) {
```
```suggestion
// computeDesiredL4PolicyMapEntries transforms the desired EndpointPolicy.L4Policy into the
// datapath-friendly format inside EndpointPolicy.PolicyMapState.
func (p *EndpointPolicy) computeDesiredL4PolicyMapEntries(identityCache cache.IdentityCache) {
```

ianvernon replied:
Documentation will be added in commit in next upload.

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -0,0 +1,124 @@
+					continue
+				}
+			}
+			policyKeys[keyFromFilter] = MapStateEntry{ProxyPort: proxyPort}
```
Is the loop logic here identical between Ingress/Egress? Should it be refactored to reduce duplication?

ianvernon replied:
Will be fixed as part of a new commit in the next upload.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -528,3 +555,154 @@ func (p *Repository) GetRulesList() *models.Policy {
+	// as list of rules which actually select this endpoint. This allows us
+	// to not have to iterate through the entire rule list multiple times and
+	// perform the matching decision again when computing policy for each
+	// protocol layer, which is quite costly in terms of performance.
```
:tada:

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -111,235 +96,14 @@ func (e *Endpoint) convertL4FilterToPolicyMapKeys(filter *policy.L4Filter, direc
-	egressPolicyEnabled := e.egressPolicyEnabled
-
-	if !ingressPolicyEnabled {
-		e.getLogger().Debug("ingress policy is disabled, which equates to allow-all; allowing all identities")
```
AFAICT we're losing this context entirely. Would be helpful to retain these logs?

ianvernon replied:
I never found the log message particularly helpful when debugging policy-related issues, which is why I removed it.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -528,3 +555,154 @@ func (p *Repository) GetRulesList() *models.Policy {
+			}
+			calculatedPolicy.PolicyMapState[keyToAdd] = MapStateEntry{}
+		}
+	}
```
These L3 bits used to be nicely contained in a separate `computeDesiredL3PolicyMapEntries()` function, it could potentially improve readability to restore that for these code hunks.

ianvernon replied:
I've factored out the allowing of all identities into its own function based off of traffic direction in a future commit. However, in the future, I am going to be unifying the calculation of policy to traverse the rules only once (as opposed to three separate times for L4, CIDR, and then L3), so I'd prefer to keep the logic for if policy is enabled in the same function here.

joestringer reviewed (approved):
Mostly I have feedback around the edges which can be cleaned up in a future PR.

Somehow we've ended up with `GetRulesMatching()` and `getMatchingRules()` with very similar implementations, right next to one another. At minimum these could use the same core implementation of rule iteration and checking.

We seem to have lost the "Policy changed" information in our logs. Was this deemed not useful, or is it an unintentional loss via refactor?

Overall I found this fairly straightforward to follow once I found the relevant code, but some pieces of functionality that were removed in the final commit, such as `ComputePolicyEnforcement()` and the `localKey`/`worldKey` were moved to another location in the codebase in a previous PR so it was a little surprising to see deletion of this code without the corresponding add. I appreciate it can be a bit tricky to refactor and retain code moves within a single PR or even single commit, but if you can keep this in mind in future then it will help speed up review a bit :-)

Commit:
policy: factor out allow all at L3 logic

Factor out code which is exactly the same for ingress and egress into a function
which allows all identities in a specified IdentityCache at ingress or egress to
reduce duplication of logic.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: factor out duplicated L4Filter conversion logic based on direction

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: add documentation for computeDesiredL4PolicyMapEntries

Signed-off by: Ian Vernon <ian@cilium.io>

tgraf reviewed (approved):
This is amazing work, great work @ianvernon!
